jetty-websocket/javax-websocket-tests/src/main/java/org/eclipse/jetty/websocket/javax/tests/NetworkFuzzer.java

```diff
@@ -73,8 +77,22 @@ public NetworkFuzzer(LocalServer server, URI wsURI, Map<String, String> requestH
             HttpFields fields = this.upgradeRequest.getHeaders();
             requestHeaders.forEach((name, value) ->
             {
-                fields.remove(name);
-                fields.put(name, value);
+                if (HttpHeader.SEC_WEBSOCKET_SUBPROTOCOL.toString().equalsIgnoreCase(name))
```

Comment:
@gregw this loop was setting `HttpFields` of the `HttpRequest` directly instead of going through the setters on `ClientUpgradeRequest`, because of this it was failing the new checks in `ClientUpgradeRequest.upgrade()`

I put in this hack to fix the `NetworkFuzzer` so that it treats subprotocol and extension headers differently. But if we want to allow direct setting of these headers bypassing the methods on `ClientUpgradeRequest` then I will probably need to reparse `extensions` and `subProtocols` from the headers before they are used for verification (then there's not much point even storing them as fields)

Reply:
At issue here is that a) I'm not sure we can prevent access to the `HttpFields` object;  b) We don't have an immutable version of `HttpFields`; c) it can be more efficient to mutate `HttpFields` directly than to use string based methods on other classes.

So I tend to agree that there is probably not much point in storing them as fields.... in fact I had thought we had already changed that in negotiation? So let's get rid of the duplicated state (field and header) if possible.